File: gridcodec/js/test_gridcodec.js
/**
 * GridCodec tests + performance. Run with: node test_gridcodec.js
 */
const gc = require('./gridcodec.js');

const { fieldIndex, fieldName, gridIndex, gridName, GridCodecMatrix } = gc;

function testHelpers() {
  console.log('  test field_index / field_name...');
  if (fieldIndex('FN') !== 5 * 18 + 13) throw new Error('fieldIndex FN');
  if (fieldName(fieldIndex('FN')) !== 'FN') throw new Error('fieldName');
  if (fieldIndex('OL') !== 14 * 18 + 11) throw new Error('fieldIndex OL');
  console.log('  OK');

  console.log('  test grid_index / grid_name...');
  if (gridName(gridIndex('FN31')) !== 'FN31') throw new Error('grid name');
  console.log('  OK');
}

function testEmptyRoundtrip() {
  console.log('  test_empty_roundtrip...');
  const m = new GridCodecMatrix();
  const { data, len } = m.encode();
  if (len <= 0 || data[0] !== 0x01 || data[1] !== 0) throw new Error('empty encode');
  const m2 = new GridCodecMatrix();
  const consumed = m2.decode(data);
  if (consumed !== len) throw new Error('decode len');
  console.log(' OK (encoded ' + len + ' bytes)');
}

function testSinglePathRoundtrip() {
  console.log('  test_single_path_roundtrip...');
  const m = new GridCodecMatrix();
  if (m.set('FN31', 'PM02') !== 0) throw new Error('set');
  const { data, len } = m.encode();
  if (len <= 0) throw new Error('encode');
  const m2 = new GridCodecMatrix();
  const consumed = m2.decode(data);
  if (consumed !== len) throw new Error('decode');
  const fromFn = m2.gcFrom('FN31');
  if (fromFn.length !== 1 || gridName(fromFn[0]) !== 'PM02') throw new Error('gc_from');
  const toPm = m2.gcTo('PM02');
  if (toPm.length !== 1 || gridName(toPm[0]) !== 'FN31') throw new Error('gc_to');
  console.log(' OK (encoded ' + len + ' bytes)');
}

function testQueryFromTo() {
  console.log('  test_query_from_to...');
  const m = new GridCodecMatrix();
  m.set('FN31', 'PM02');
  m.set('FN31', 'PM03');
  m.set('JO22', 'FN31');
  const fromFn = m.gcFrom('FN31');
  if (fromFn.length !== 2) throw new Error('from count');
  const names = fromFn.map(i => gridName(i));
  if (!names.includes('PM02') || !names.includes('PM03')) throw new Error('from names');
  const toFn = m.gcTo('FN31');
  if (toFn.length !== 1 || gridName(toFn[0]) !== 'JO22') throw new Error('to');
  console.log(' OK');
}

function testDedup() {
  console.log('  test_dedup...');
  const m = new GridCodecMatrix();
  m.set('OL72', 'FN31');
  m.set('OL72', 'FN31');
  const { data, len } = m.encode();
  const m2 = new GridCodecMatrix();
  m2.decode(data);
  const fromOl = m2.gcFrom('OL72');
  if (fromOl.length !== 1 || gridName(fromOl[0]) !== 'FN31') throw new Error('dedup');
  console.log(' OK');
}

function testCInterop() {
  console.log('  test_c_interop (decode C payload)...');
  const payload = new Uint8Array([
    0x01, 0x00, 0x20, 0x02, 0x00, 0x80, 0x01, 0x02, 0x08, 0x00, 0x0c, 0x09,
    0x06, 0x06
  ]);
  const m = new GridCodecMatrix();
  const consumed = m.decode(payload);
  if (consumed !== payload.length) throw new Error('consumed');
  const fromFn = m.gcFrom('FN');
  if (fromFn.length !== 1 || fieldName(fromFn[0]) !== 'PM') throw new Error('from FN');
  const toFn = m.gcTo('FN');
  if (toFn.length !== 1 || fieldName(toFn[0]) !== 'JO') throw new Error('to FN');
  console.log(' OK');
}

function testRealistic(nPaths) {
  console.log('  test_realistic_roundtrip (' + nPaths + ' paths)...');
  const m = new GridCodecMatrix();
  const grids = [];
  for (let i = 0; i < nPaths * 2; i++) {
    const fi = Math.floor(Math.random() * gc.GC_FIELDS);
    const si = Math.floor(Math.random() * gc.GC_SQUARES);
    const lon = Math.floor(si / 10), lat = si % 10;
    grids.push(fieldName(fi) + lon + lat);
  }
  for (let i = 0; i < nPaths * 2 - 1; i += 2) m.set(grids[i], grids[i + 1]);
  const t0 = performance.now();
  const { data, len } = m.encode();
  const t1 = performance.now();
  const m2 = new GridCodecMatrix();
  const consumed = m2.decode(data);
  const t2 = performance.now();
  if (consumed !== len) throw new Error('decode len');
  let verified = 0;
  for (let i = 0; i < nPaths * 2 - 1; i += 2) {
    const out = m2.gcFrom(grids[i]);
    if (out.some(idx => gridName(idx) === grids[i + 1])) verified++;
  }
  if (verified !== nPaths) throw new Error('verified ' + verified);
  console.log(' OK');
  console.log('    Encoded: ' + len + ' bytes, Encode: ' + (t1 - t0).toFixed(2) + ' ms, Decode: ' + (t2 - t1).toFixed(2) + ' ms, Verified: ' + verified);
}

function testPerformance() {
  console.log('  [Performance] 500 / 5000 / 2000 paths...');
  for (const n of [500, 5000, 2000]) {
    const m = new GridCodecMatrix();
    for (let i = 0; i < n; i++) {
      const a = [Math.floor(Math.random() * gc.GC_FIELDS), Math.floor(Math.random() * gc.GC_SQUARES)];
      const b = [Math.floor(Math.random() * gc.GC_FIELDS), Math.floor(Math.random() * gc.GC_SQUARES)];
      const fromG = fieldName(a[0]) + Math.floor(a[1] / 10) + (a[1] % 10);
      const toG = fieldName(b[0]) + Math.floor(b[1] / 10) + (b[1] % 10);
      m.set(fromG, toG);
    }
    const t0 = performance.now();
    const { data, len } = m.encode();
    const t1 = performance.now();
    const m2 = new GridCodecMatrix();
    m2.decode(data);
    const t2 = performance.now();
    console.log('    ' + n + ' paths: encoded ' + len + ' bytes, encode ' + (t1 - t0).toFixed(2) + ' ms, decode ' + (t2 - t1).toFixed(2) + ' ms');
  }
}

function main() {
  console.log('=== GridCodec JavaScript Test Suite ===\n');
  Math.seedrandom ? Math.seedrandom(42) : null;
  console.log('[Helpers]');
  testHelpers();
  console.log('\n[Round-trip]');
  testEmptyRoundtrip();
  testSinglePathRoundtrip();
  console.log('\n[Query]');
  testQueryFromTo();
  testDedup();
  console.log('\n[Realistic]');
  testRealistic(500);
  testRealistic(5000);
  testRealistic(2000);
  console.log('\n[C interop]');
  testCInterop();
  console.log('\n[Performance]');
  testPerformance();
  console.log('\n=== All tests passed ===\n');
}

main();

File: gridcodec/python/test_gridcodec.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
"""Tests and performance for gridcodec (full-featured Python)."""
from __future__ import print_function
import os
import sys
import time

# Allow running from python/ so that "gridcodec" package is found
sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))
from gridcodec import (
    GC_FIELDS, GC_SQUARES, GC_GRIDS,
    field_index, field_name, grid_index, grid_name,
    grid_to_field, grid_to_square,
    GridCodecMatrix,
    GC_ERR_INVALID, GC_ERR_OVERFLOW, GC_ERR_FORMAT,
)


def test_helpers():
    print("  test field_index / field_name...", end="")
    assert field_index("FN") == 5 * 18 + 13
    assert field_index("fn") == field_index("FN")
    assert field_index("OL") == 14 * 18 + 11
    assert field_name(field_index("FN")) == "FN"
    assert field_index("RR") >= 0  # last valid field
    assert field_index("X") == -1
    assert field_index("") == -1
    print(" OK")

    print("  test grid_index / grid_name...", end="")
    assert grid_index("FN31") == field_index("FN") * 100 + 31
    assert grid_name(grid_index("FN31")) == "FN31"
    assert grid_index("OL72") >= 0
    assert grid_to_field(grid_index("FN31")) == field_index("FN")
    assert grid_to_square(grid_index("FN31")) == 31
    print(" OK")


def test_empty_roundtrip():
    print("  test_empty_roundtrip...", end="")
    m = GridCodecMatrix()
    data, n = m.encode()
    assert n > 0
    assert data[0] == 0x01 and data[1] == 0
    m2 = GridCodecMatrix()
    consumed = m2.decode(data)
    assert consumed == n
    print(" OK (encoded %d bytes)" % n)


def test_single_path_roundtrip():
    print("  test_single_path_roundtrip...", end="")
    m = GridCodecMatrix()
    r = m.set("FN31", "PM02")
    assert r == 0
    data, n = m.encode()
    assert n > 0
    m2 = GridCodecMatrix()
    consumed = m2.decode(data)
    assert consumed == n
    from_fn = m2.gc_from("FN31")
    assert len(from_fn) == 1
    assert grid_name(from_fn[0]) == "PM02"
    to_pm = m2.gc_to("PM02")
    assert len(to_pm) == 1
    assert grid_name(to_pm[0]) == "FN31"
    print(" OK (encoded %d bytes)" % n)


def test_layer1_only():
    print("  test_layer1_only (two paths, L1+L2)...", end="")
    m = GridCodecMatrix()
    m.set("FN31", "PM02")
    m.set("JO22", "FN31")
    data, n = m.encode()
    assert data[1] & 1  # has_layer2
    m2 = GridCodecMatrix()
    consumed = m2.decode(data)
    assert consumed == n
    assert len(m2.gc_from("FN")) == 1
    assert field_name(m2.gc_from("FN")[0]) == "PM"
    print(" OK (encoded %d bytes)" % n)


def test_query_from_to():
    print("  test_query_from_to...", end="")
    m = GridCodecMatrix()
    m.set("FN31", "PM02")
    m.set("FN31", "PM03")
    m.set("JO22", "FN31")
    from_fn = m.gc_from("FN31")
    assert len(from_fn) == 2
    assert set(grid_name(x) for x in from_fn) == {"PM02", "PM03"}
    to_fn = m.gc_to("FN31")
    assert len(to_fn) == 1
    assert grid_name(to_fn[0]) == "JO22"
    from_jo = m.gc_from("JO22")
    assert len(from_jo) == 1
    assert grid_name(from_jo[0]) == "FN31"
    # 2-char query
    from_f = m.gc_from("FN")
    assert len(from_f) == 1 and field_name(from_f[0]) == "PM"
    print(" OK")


def test_dedup():
    print("  test_dedup (same pair twice)...", end="")
    m = GridCodecMatrix()
    m.set("OL72", "FN31")
    m.set("OL72", "FN31")
    data, n = m.encode()
    m2 = GridCodecMatrix()
    m2.decode(data)
    from_ol = m2.gc_from("OL72")
    assert len(from_ol) == 1 and grid_name(from_ol[0]) == "FN31"
    print(" OK")


def test_realistic_roundtrip(n_paths=500):
    print("  test_realistic_roundtrip (%d paths)..." % n_paths, end="")
    import random
    random.seed(42)
    m = GridCodecMatrix()
    grids = []
    for _ in range(n_paths * 2):
        fi = random.randint(0, GC_FIELDS - 1)
        si = random.randint(0, GC_SQUARES - 1)
        grids.append(field_name(fi) + "%d%d" % (si // 10, si % 10))
    for i in range(0, len(grids) - 1, 2):
        m.set(grids[i], grids[i + 1])
    t0 = time.perf_counter()
    data, enc_len = m.encode()
    t1 = time.perf_counter()
    m2 = GridCodecMatrix()
    t2 = time.perf_counter()
    consumed = m2.decode(data)
    t3 = time.perf_counter()
    assert consumed == enc_len
    verified = 0
    for i in range(0, len(grids) - 1, 2):
        out = m2.gc_from(grids[i])
        if any(grid_name(x) == grids[i + 1] for x in out):
            verified += 1
    assert verified == n_paths
    enc_ms = (t1 - t0) * 1000
    dec_ms = (t3 - t2) * 1000
    print(" OK")
    print("    Encoded: %d bytes, Encode: %.2f ms, Decode: %.2f ms, Verified: %d" % (
        enc_len, enc_ms, dec_ms, verified))


def test_performance():
    print("  [Performance] 500 / 5000 / 2000 paths...")
    for n in (500, 5000, 2000):
        import random
        random.seed(123)
        m = GridCodecMatrix()
        for _ in range(n):
            a = (random.randint(0, GC_FIELDS - 1), random.randint(0, GC_SQUARES - 1))
            b = (random.randint(0, GC_FIELDS - 1), random.randint(0, GC_SQUARES - 1))
            from_g = field_name(a[0]) + "%d%d" % (a[1] // 10, a[1] % 10)
            to_g = field_name(b[0]) + "%d%d" % (b[1] // 10, b[1] % 10)
            m.set(from_g, to_g)
        t0 = time.perf_counter()
        data, enc_len = m.encode()
        t1 = time.perf_counter()
        m2 = GridCodecMatrix()
        consumed = m2.decode(data)
        t2 = time.perf_counter()
        enc_ms = (t1 - t0) * 1000
        dec_ms = (t2 - t1) * 1000
        print("    %d paths: encoded %d bytes, encode %.2f ms, decode %.2f ms" % (
            n, enc_len, enc_ms, dec_ms))


def test_c_interop():
    """Decode a payload produced by C (from test_embedded.c payload_l1_only)."""
    print("  test_c_interop (decode C payload)...", end="")
    payload_l1_only = bytes([
        0x01, 0x00, 0x20, 0x02, 0x00, 0x80, 0x01, 0x02, 0x08, 0x00, 0x0c, 0x09,
        0x06, 0x06
    ])
    m = GridCodecMatrix()
    consumed = m.decode(payload_l1_only)
    assert consumed == len(payload_l1_only)
    from_fn = m.gc_from("FN")
    assert len(from_fn) == 1 and field_name(from_fn[0]) == "PM"
    to_fn = m.gc_to("FN")
    assert len(to_fn) == 1 and field_name(to_fn[0]) == "JO"
    print(" OK")


def main():
    print("=== GridCodec Python Test Suite ===\n")
    print("[Helpers]")
    test_helpers()
    print("\n[Round-trip]")
    test_empty_roundtrip()
    test_single_path_roundtrip()
    test_layer1_only()
    print("\n[Query]")
    test_query_from_to()
    test_dedup()
    print("\n[Realistic]")
    test_realistic_roundtrip(500)
    test_realistic_roundtrip(5000)
    test_realistic_roundtrip(2000)  # 20k is very slow in pure Python
    print("\n[C interop]")
    test_c_interop()
    print("\n[Performance]")
    test_performance()
    print("\n=== All tests passed ===\n")
    return 0


if __name__ == "__main__":
    sys.exit(main())
